---
title: "Additional_annot_opts"
author: "Tania Guerrero"
date: "12/17/2019"
output: html_document
---

This script it's full of failed attempts to add annotation with different tools and packages. I kept it bc for the experienced eye it might have useful information. 

*Prepare annotation data*
Annotation data should be an object of GRanges. Same as import peaks, we use the method toGRanges, which can return an object of GRanges, to represent the annotation data. An annotation data be constructed from not only BED, GFF or user defined readable text files, but also EnsDb or TxDb object, by calling the toGRanges method. Please type ?toGRanges for more information.
```{r}
library(ChIPpeakAnno)
library(rtracklayer)
library(MEDIPS)

rawGFF <- readGFF("/data/fg2/guerrero/cheetah_methylation/MeDIPS/data/genome/refGenome/Acijub_2_GCF_003709585.1_genomic.gff") ## difference is a dataframe and with import it creates already a GRanges object

gff <- rtracklayer::import("/data/fg2/guerrero/cheetah_methylation/MeDIPS/data/genome/refGenome/Acijub_2_GCF_003709585.1_genomic.gff")

annotData <- toGRanges("/data/fg2/guerrero/cheetah_methylation/MeDIPS/data/genome/refGenome/Acijub_2_GCF_003709585.1_genomic.gff", format= "GFF") ### OKAY THIS IS NOT WORKING...BUT I guess I can use the gff since that's already a GRanges object---FIGURE OUT WHY IS NOT WORKING... MAYBE THAT'S WHY IS NOT DOING THE OVERLAPPING PROPERLY.... :s 
annotData[1:2] #???

# so let's convert the results table to GRanges
#meth_data_frame <- data.frame(matrix(unlist(methList_test), nrow = length(methList_test), byrow = T))

#I am taking object 4 from the MTCmethList cuz is the one with the most astringent p-val, just for the quick test...
MTCp-val_0.05 <- (MTCmethList[[4]])

TEST <- MTCmethList[[4]]

# Apparently  I need to change the names...so that it matches with GRanges format
colnames(MTCp-val_0.05) <- c("seqname", "start", "end", "ID", "length", "numSigWin", "minP", "cpgNum", "cpgDensity")
# methList_test <- toGRanges(meth_data_frame)### this one is not working..let's try makeGRangesFromDataFrame
test_GRanges <- makeGRangesFromDataFrame(MTCp-val_0.05)

#I DUNNO ...how to convert my methList into GRanges object... 
#singleGRange <- GRanges(as.data.frame(GRangesList(methList)))
#methListGRanges <- toGRanges(methList) #: unable to find an inherited method for function 'toGRanges' for signature '"list"': unable to find an inherited method for function 'toGRanges' for signature '"list"'


##Annotate "peaks" with annotatePeakInBatch
annot_test = annotatePeakInBatch(test_GRanges, AnnotationData=gff_acijub)
annot_test = addGeneIDs(annot_test, gff, IDs2Add = "gene") ### ????

gff@elementMetadata@listData[["gene"]]

#Using M. Skinner's approach
 overlaps <- findOverlaps(test_GRanges, gff)
    # add annotation to dmrList
    dmrList_test <- cbind(test_GRanges, annotation = NA)
    dmrList_test$annotation[queryHits(overlaps)] <- as.character(gff$group[subjectHits(overlaps)])
###is somehow missing the data inside metadata... how do I fetch that???
    
## MEDIPS annotation
gff_mart = MEDIPS.getAnnotation(host= "www.ncbi.nlm.nih.gov", dataset = gff, annotation = (gff@elementMetadata@listData[["gene"]]))
annot_test_b = MEDIPS.setAnnotation(regions = test_GRanges, annotation = gff)

?MEDIPS.getAnnotation


```

```{r setup, include=FALSE}
library(BiocManager)
#BiocManager::install("Rsubread")

#library(devtools)
#devtools::install_github("torkian/subread-1.6.1") # esto es una puta mierda!!! 
```

```{r}
library(Rsubread)
# Create annot file from a GRanges object which I might get out of the other package (see following chunk)
createAnnotationFile(gff) #value: A data frame with five columns ###Your GRanges elementMetadata must contain a column called "id".
# Error in createAnnotationFile(gff) : Aborting. namedGeneID,Chr,Start,EndandStrand.

createAnnotationFile(rawGFF) #To create Rsubread annotations from a TxDb or FeatureDb,
#extract a GRanges using exons(), transcripts(), or features().
#Error in createAnnotationFile(rawGFF) : Aborting.

write.Rsubread(gff)

#at some point give the proper name
#when having the package loaded check for list of preloaded annotations

#Built-in annotations can be accessed via function 
featureCounts #to see the SAF format 
getInBuiltAnnotation()
 
aciJub1 = gff

?promoterRegions
promoterRegions()
promoterRegions(annotation=aciJub1, upstream=3000L, downstream=2000L) 
###not working... below the source of the fuction... how to choose values...?

```

```{r}

promoterRegions <- function(annotation=aciJub1, upstream=3000L, downstream=2000L)   
#	Create a SAF data.frame of genewise promoter regions
#	Gordon Smyth
#	24 April 2017
{
#	annotation can be a SAF format data.frame or can be the name of a genome with built-in annotation
	if(is.character(annotation)) annotation <- getInBuiltAnnotation(annotation)

#	Remove unassembled contigs
	N <- grep("^N",annotation$Chr)
	annotation <- annotation[-N,]

#	Check upstream and downstream limits
	upstream <- as.integer(upstream[1])
	if(upstream < 0L) upstream <- 0L
	downstream <- as.integer(downstream[1])
	if(downstream < 0L) downstream <- 0L

#	Combine Chr and GeneID
	annotation$ChrGeneID <- paste(annotation$Chr,annotation$GeneID,sep=".")

#	Get start of each gene
	o <- order(annotation$ChrGeneID,annotation$Start)
	anno.start <- annotation[o,]
	isdup <- duplicated(anno.start$ChrGeneID)
	exon.first <- which(!isdup)
	anno.start <- anno.start[exon.first,]

#	Get end of each gene
	o <- order(annotation$ChrGeneID,annotation$End)
	anno.end <- annotation[o,]
	exon.last <- c(exon.first[-1]-1L,nrow(annotation))
	anno.end <- anno.end[exon.last,]

#	Assemble start and end into one data.frame
	ann.gene <- anno.start
	ann.gene$End <- anno.end$End

#	Compute promoter region for + strand
	prom.start <- pmax(ann.gene$Start-upstream,1L)
	prom.end <- pmin(ann.gene$Start+downstream,ann.gene$End)

#	Compute promoter region for - strand
	neg <- ann.gene$Strand=="-"
	prom.start[neg] <- pmax(ann.gene$End[neg]-downstream,ann.gene$Start[neg])
	prom.end[neg] <- ann.gene$End[neg]+upstream

#	Output
	ann.gene$Start <- prom.start
	ann.gene$End <- prom.end
	ann.gene$ChrGeneID <- NULL
	o <- order(ann.gene$Chr,ann.gene$Start)
	ann.gene[o,]
}
```

#Option using toGRanges
Burrowed from Chip-seq experiments
```{r}
BiocManager::install("ChIPpeakAnno", version = "3.8")
install.packages("MASS")
install.packages("Matrix")
install.packages("mgcv")
install.packages("survival")
library(ChIPpeakAnno)

```

### Annotation trials 28 Aug
```{r}
BiocManager::install("AnnotationHub")
library(AnnotationHub)
hub <- AnnotationHub()
hub <- subset(hub, hub$species=="Acinonyx jubatus")
length(hub)
#of course cheetah is not there ¬¬ how naive of me haha
```


```{r}
#BiocManager::install("RIPSeeker")
library(RIPSeeker)

if(interactive()) { 
  outfile <- paste(getwd(), "/MTC_p-val_0.05_promoters.txt", sep = "/")
exportGRanges(loc_acijub_prom, outfile = outfile, exportFormat = "txt")
}

?exportGRanges
```


##This does not help either ... :/
```{r}
library(biomaRt)
install.packages("biomartr")
library(biomartr)
biomartr::getGenome(db="refseq", organism = "Acinonyx jubatus", path = file.path("_ncbi_downloads", "genomes"))
```



